tuf/tuf.go

```diff
@@ -143,13 +143,16 @@ func (tr *Repo) RemoveBaseKeys(role string, keyIDs ...string) error {
 		}
 	}
 
-	// remove keys no longer in use by any roles
-	for k := range toDelete {
-		delete(tr.Root.Signed.Keys, k)
-		// remove the signing key from the cryptoservice if it
-		// isn't a root key. Root keys must be kept for rotation
-		// signing
-		if role != data.CanonicalRootRole {
+	// Remove keys no longer in use by any roles, except for root keys.
+	// Root private keys must be kept in tr.cryptoService to be able to sign
+	// for rotation, and root certificates must be kept in tr.Root.SignedKeys
+	// because we are not necessarily storing them elsewhere (tuf.Repo does not
+	// depend on certs.Manager, that is an upper layer), and without storing
+	// the certificates in their x509 form we are not able to do the
+	// util.CanonicalKeyID conversion.
+	if role != data.CanonicalRootRole {
+		for k := range toDelete {
+			delete(tr.Root.Signed.Keys, k)
```

Comment:
> I would greatly prefer to use the previous root.json's signature list to decide which old keys to sign with.

Yes, that’s what this PR is already doing (`oldKeyIDs` in `Repo.SignRoot`). But the old signature list only contains X509-based key IDs, which are sufficient neither to regenerate the X509 certificates nor to even locate the non-X509 private key which should be used for signing.

> Using that model, ceasing to sign with an old key (i.e. because it is years old and considered no longer relevant to rotations) is as simple as deleting if from the keystore

This also already works that way with this PR: The `oldKeyIDs` mentioned above eventually propagate to `signed.Sign`’s `optionalKeys`; if the underlying private key for any of the `optionalKeys` can not be loaded, this key ID will be silently skipped, and none of the future signatures of that `root.json` will use that key any more. The related X.509 certificate will be kept in `Signed.Keys` forever, but having unused public keys in there does not hurt anything AFAICS.